homeassistant/components/rainmachine/__init__.py

```diff
@@ -288,15 +288,42 @@ async def async_init_coordinator(
 
     entry.async_on_unload(entry.add_update_listener(async_reload_entry))
 
-    async def async_pause_watering(call: ServiceCall) -> None:
+    def call_with_controller(update_programs_and_zones: bool = True) -> Callable:
+        """Hydrate a service call with the appropriate controller."""
+
+        def decorator(func: Callable) -> Callable[..., Awaitable]:
+            """Define the decorator."""
+
+            @wraps(func)
+            async def wrapper(call: ServiceCall) -> None:
+                """Wrap the service function."""
+                entry = async_get_entry_for_service_call(hass, call)
+                data: RainMachineData = hass.data[DOMAIN][entry.entry_id]
```

Comment:
Isn't this exactly the same logic to get the controller, just in a different spot?

Reply:
That part is – the primary difference is a few lines down:

```
if update_programs_and_zones:
    await async_update_programs_and_zones(hass, entry)
```

That entry isn't always the correct one.